### DISCUSSION TITLE
Migrating from Flash Builder to IntelliJ

### DISCUSSION THREAD

**Discussion by raresn** [airsdk/Adobe-Runtime-Support#1999] [Guides & Tutorials]
OK, so we decided to give IntelliJ another shot :)
First of all - we are complete beginners with IntelliJ, so I hope the info we will gather here will help someone at some point 
We installed IntelliJ
- installed the Flash/Flex plugin
- imported the flash builder project
- changed the AIR SDK path to the right place (from Exteral libraries in the "Project" panel)

Then we just hit Run and we manage to export a SWF.

We use quite a few of the ANEs from distriqt, and in Flash Builder we had the option define "Compiler Arguments"
![image](https://user-images.githubusercontent.com/1768694/176701947-a79a08b3-48e7-4153-957d-d439d828329c.png)
This means, that we changed those DESKTOP and MOBILE variable from true to false before compiling in order to get the right code in for Mobile, or Desktop
Something like:
```
COMPILE::MOBILE 
{
      FirebaseFCM2.unsubscribe();
}
COMPILE::DESKTOP {
      Api.unregisterDevice();
}
```
Is there a way to do this in IntelliJ as well, @marchbold , what do you recommend?

Would be cool if anyone has tips and tricks on how to debug/compile to share them here :)

**Comment by raresn** [comment]
Found out how we can add the compiler arguments in IntelliJ for anyone who needs this:
![image](https://user-images.githubusercontent.com/1768694/176705094-444e2e58-7315-4e18-a3e8-2f2f927248a1.png)

**Reply by marchbold** [reply]
Glad to see you found them 👍

**Reply by raresn** [reply]
@marchbold, so moving down the error lane... i remember why we were unable to switch to IntelliJ :) Maybe with your help we will manage to move forward.

We have a class, with a method:
```
		private function addMethodsTo(cls:Class, methods):void
		{
			for (var name:String in methods){
				cls.prototype[name] = methods[name];
				cls.prototype.setPropertyIsEnumerable(name, false);
			}
		}
```
Inside the class constructor we have:
```
addMethodsTo(String, 
{
      toStringZeroPrefix: function():String
       {
           return String("00" + this).substr(-2);
       }
});

addMethodsTo(Date, 
{
      toYMD: function():String {
           return this.fullYear.toString() + this.month.toString().toStringZeroPrefix() + this.date.toString().toStringZeroPrefix();
      }		
});
```

Which IntelliJ throws a problem; Unresolved function or method toStringZeroPrefix() ONLY for "this.month.toString().toStringZeroPrefix()"
this.date.toString().toStringZeroPrefix(); doesn't throw a problem, but it is greyed out.
![image](https://user-images.githubusercontent.com/1768694/176856427-02b02b13-ef2d-45de-bdef-4ca12de31a8c.png)


Any idea what we can do?

**Reply by ajwfrost** [reply]
Hi @raresn  - I've not tried that out yet but just looking at the error:
> Unresolved function or method toStringZeroPrefix() ONLY for "this.month.toString().toStringZeroPrefix()"
>  this.date.toString().toStringZeroPrefix(); doesn't throw a problem..

I'm wondering whether the problem is that `this.month.toString()` isn't actually returning a `String` object... what are the types of those `month` and `date` properties?

Potentially it could work with casting e.g. `(this.month.toString() as String).toStringZeroPrefix()`?

Ultimately though: IntelliJ would be likely be using the new AS3 compiler, and Flash Builder could be using old one (to have parity with Flex) so there are a few differences. If you could set up IntelliJ to use an SDK that's made up of the Flex SDK with the "AIR for Flex Developers" SDK added over the top, it may solve it (or point IntelliJ at the existing compiler/SDK that Flash Builder is using?)

thanks

**Reply by raresn** [reply]
Hi, @ajwfrost
> "_I'm wondering whether the problem is that this.month.toString() isn't actually returning a String object... what are the types of those month and date properties?_"

They are the properties of a simple Date object :)
We created a static version here toStringZeroPrefix that i can use - probably will need to create a string util class with this just to be able to work around this.
Thing is, is that it builds without errors... 
I would very much like to avoid doing more SDK overlays :)

Check this out:
![image](https://user-images.githubusercontent.com/1768694/176884276-7d3713a6-4643-425e-aa5b-1676c056956e.png)

Sincer we are talking about the Date object, we would benefit from a few extra functions like we have here:
https://dev.mysql.com/doc/refman/8.0/en/date-and-time-functions.html
Like:
TO_DAYS
YEARWEEK
WEEK
WEEKDAY - we have this but 0 Monday would help us most, so that it returns Monday as 0 without us doing math to get it.
(Not to mention the TimeZone options that we discussed in an email and also mentioned here: [https://github.com/airsdk/Adobe-Runtime-Support/issues/401](url) )

-----
The next challenge would be to launch the dekstop version of our app in the AIR Simulator, and put some breakpoints around to debug.

**Comment by M4rkY4** [comment]
@raresn I believe you can't explain compiler at compile time that later at runtime somewhere within some condition you might add a new method there. Maybe, there's some compiler argument for that or way to disable some strict mode.

You can always use `myString["toStringZeroPrefix"]();`
or

```
class Consts
{ static public const TO_STRING_ZERO_PREFIX = "toStringZeroPrefix"; }

myString[Consts.TO_STRING_ZERO_PREFIX]();
```
Looks ugly, but works.

Red text is IDEA inspections, you can disable them, but this will not fix compiler issue.

![Screenshot 2022-07-01 at 14 08 15](https://user-images.githubusercontent.com/93158631/176883520-a7c234e3-448f-41f9-9c9f-7764afe8a11f.png)

![Screenshot 2022-07-01 at 14 08 58](https://user-images.githubusercontent.com/93158631/176883540-f53f7f4b-107f-4b90-a733-620746576f3a.png)

**Reply by raresn** [reply]
@M4rkY4 Nice!
What do you do in order to debug in IDEA?
In flash builder we used to choose to run in the AIR Simulator, and hit f12 :) Any tips on doing this here?
AND would also need to be able to debug on device. Wondering if there are some tutorials out there, but if you have time to explain these things, I thing it would be super good for other people as well.

**Reply by M4rkY4** [reply]
Nothing special,

just leaving checkbox checked:
![image](https://user-images.githubusercontent.com/93158631/176891859-86aee66d-da83-46a7-b61c-65dba20c1762.png)


creating run configurations for my needs
![image](https://user-images.githubusercontent.com/93158631/176893387-8e247cb7-8f4b-4fa3-8f76-56d15d245eb8.png)
(top right checkbox useful for saving that configuration files and copying to other projects)

selecting one of them from list (when I need to test on another screen proportions) and a bug button to debug :)
![image](https://user-images.githubusercontent.com/93158631/176890690-8d45a20d-3dce-478c-841a-4866d3622d0a.png)
I've named with "e" emulators with different resolutions for my situations I often test and with "s" builds running on devices via cable.

Flash App for debugging with rebuilding project (99% of time), Flash Remote Debug to debug existing build on device without rebuilding (code changes will not be applied, useful for searching how to reproduce an existing bug)
![image](https://user-images.githubusercontent.com/93158631/176891692-4fbf419d-77e9-42bf-ac7b-628459255544.png)
![image](https://user-images.githubusercontent.com/93158631/176895614-245d5387-8242-4a4d-a5e7-a9491ee2af2f.png)


The complicated part is to properly create configuration XML, but it's not specific for IDEA.

Sadly, recently Android with AAB became way too complicated to debug and configure builds, at least for me, I'm using specially configured Ant for builds.

**Reply by raresn** [reply]
We use AAB for production only, so we created a custom BAT that will just package things in the desired way.

Now, when i try to run the app in debug for device i get this:
Executing pre-compile tasks...
Loading Ant configuration...
Running Ant tasks...
Running 'before' tasks
Checking sources
Compiling Agenda (module AgendaOld)...
[Agenda (module AgendaOld)]: d:\jdk-17.0.2\bin\java.exe -Dapplication.home=D:\AdobeAIRSDK -Dfile.encoding=UTF-8 -Djava.awt.headless=true -Duser.language=en -Duser.region=en -Xmx512m -classpath D:\AdobeAIRSDK\lib\compiler.jar com.adobe.flash.compiler.clients.MXMLC -load-config=C:\Users\FxStudio\AppData\Local\JetBrains\IntelliJIdea2022.1\compile-server\agendaold_f9486e48\_temp_\IntelliJ_IDEA\idea-5793A1F3-84F4C647.xml -define+=COMPILE::DESKTOP,true -define+=COMPILE::MOBILE,false

Waits a bit, then BOOM:
[Agenda (module AgendaOld)]: AIR descriptor file not found: [PATH_TO_PROJECT]/ANDROID_AAB.XML

It actually deleted the descriptor file :) That file was there, and it just deleted it and then said it isn't there.
Anyone experienced this before?
https://intellij-support.jetbrains.com/hc/en-us/community/posts/206813985-Descriptor-for-Adobe-AIR-Mobile-App-Gets-Deleted-by-IntelliJ-when-Compiling-
I see someone here mentioned this as well.

**PS - I moved the descriptor file to another folder, and pointed there - that seems to do the trick....**
** So right now i have the XML descriptor in a separate folder and all the libraries + extensions in a separate folder **

**Reply by raresn** [reply]
![image](https://user-images.githubusercontent.com/1768694/176910188-d1147d8a-a634-4ca1-8eae-ddfe047fcd54.png)
After moving the XML file from the bin-debug folder .... I get the above. Wasn't it supposed to start the app + debug session?

**Reply by raresn** [reply]
For those of you who struggle with this, you need to go to Run-> Edit configurations -> Press the + -> and choose Flash Remote Debug :)
Thanks @M4rkY4

**Comment by Adrian-S** [comment]
![image](https://user-images.githubusercontent.com/1580105/178688312-cf9b8c1b-b506-413b-b0af-5b49f3e1e5b2.png)

I think IntelliJ is excelent, but failed to a simple type error. Error is only visible on project level when I build the project.

Am I doing something wrong?

**Reply by raresn** [reply]
@M4rkY4 any ideas on how to proceed here?

**Reply by M4rkY4** [reply]
Just created new file, it warns me right away, no run/debug/build commands used

![image](https://user-images.githubusercontent.com/93158631/178709849-69be36f1-1836-4f2a-95aa-e0c746737d2b.png)

I assume the project is not fully set up,
File -> Project Structure -> Project Settings -> tabs "Project -> Project SDK" & "Modules -> xxxx -> Dependencies -> Flex/AIR SDK" your Air/Flex SDK should be set there
And on bottom right of IDEA your module xxxx configuration should be selected for highlighting

**Reply by Adrian-S** [reply]
![image](https://user-images.githubusercontent.com/1580105/178713296-add91e9a-a6e0-412c-a5d4-46a4a8e60002.png)
This MUST be set to "All Problems". It was highliting too much and I changed it. And now that I've put it back I have rainbows on the scroll bar and don't know how to fine tune it for maximum visibility and productivity.

As I was saying. IntelliJ is better, but it's unnecesary too complex and I can't find what I need when I need it to the point of wasteing my time searching for color scheme options.

**Reply by raresn** [reply]
@ajwfrost this is a thing that really hits in our testing time - each time we debug the app, the compile time is around **15-20 seconds**. In flash builder this was super fast (compiling the same code). 
Do we need to have any setting checked for this to go faster in Intellij maybe?

**Reply by Devin-Licastro** [reply]
> @ajwfrost this is a thing that really hits in our testing time - each time we debug the app, the compile time is around **15-20 seconds**. In flash builder this was super fast (compiling the same code). Do we need to have any setting checked for this to go faster in Intellij maybe?

Hey raresn, did you ever find a way to speed up IntelliJ compile times?

**Reply by 2jfw** [reply]
> Hey raresn, did you ever find a way to speed up IntelliJ compile times?

I think this is due to the fact that IntelliJ does not support incremental compiling; this means, that IntelliJ does compile all used classes (except SWCs afaik) whenever you compile. Flash Builder did only compile the part where you actually made changes (and of course affected dependencies in other files etc) . This results in much faster compilation.

It's really sad that IntelliJ does not support this and also that they stopped even giving minor support to ActionScript.
There are some other things which would make IntelliJ much better regarding compile time and which could be solved (probably quite easily) by Jetbrains. E.g. the fact that IntelliJ uses to extract ANEs all the time when compiling, though it could use some caching for these (and optionally do some md5 checks to verfiy this).

Pretty sure that Harman cannot do anything here as it's solely up to Jetbrains.
